src/orquestra/quantum/circuits/_itertools.py

```diff
@@ -0,0 +1,179 @@
+from itertools import islice
+from math import ceil
+from typing import Sequence, Tuple, TypeVar, Iterable, Dict
+
+from orquestra.quantum.measurements import Measurements
+
+T = TypeVar("T")
+
+
+def _iterate_in_batches(items: Iterable[T], batch_size: int) -> Iterable[Tuple[T]]:
+    it = iter(items)
+    while chunk := tuple(islice(it, batch_size)):
+        yield chunk
+
+
+def split_into_batches(
+    circuits: Sequence[T],
+    n_samples_per_circuit: Sequence[int],
+    max_batch_size: int,
+) -> Iterable[Tuple[Sequence[T], int]]:
+    """Split sequence of circuits and corresponding sample sizes into batches.
+
+    The circuits are split into batches of length `max_batch_size`, except
+    possibly the last batch that can be smaller. The number of samples
+    for each batch is a maximum number of samples requested for the circuits
+    in the chunk. To illustrate, consider the following input:
+
+    - circuits=[c1, c2, c3, c4, c5]
+    - n_samples_per_circuit=[n1, n2, n3, n4, n5]
+    - max_batch_size=2
+
+    Then the returned iterable contains the following tuples:
+
+    ((c1, c2), max([n1, n2]))
+    ((c3, c4), max([n3, n4]))
+    ((c5,), n5)
+
+    Args:
+        circuits: a sequence of circuits to be batched. Please note that
+          the exact type of the circuits does not matter, and hence
+          this function works for native Orquestra circuits as well as
+          circuits from other libraries/SDKs.
+        n_samples_per_circuit: a sequence of numbers of length equal to
+          len(circuits), s.t. n_samples_per_circuit[i] is number of
+          samples requested for circuits[i].
+        max_batch_size: maximum allowable size of batch. Typically it comes
+          from specific runner's restrictions.
+    Returns:
+        An iterable yielding tuples of the forms (circuits_chunk, n_samples)
+        satisfying the following constraints:
+        - len(circuits_chunk) <= max_batch_size
+        - n_samples is large enough to accommodate sample sizes requested
+          for each circuit in circuits_chunk
+    Raises:
+        ValueError: if input sequences are not of equal length, or if
+          max_batch_size is not positive.
+    """
+    if len(circuits) != len(n_samples_per_circuit):
+        raise ValueError(
+            "Mismatched lengths of `circuits` and `n_samples_per_circuit: "
+            f"({len(circuits)} and {len(n_samples_per_circuit)} respectively)."
+            "Both sequences need to have the same length"
+        )
+
+    if max_batch_size <= 0:
+        raise ValueError(
+            "Max circuits per batch has to be positive, got " f"{max_batch_size}"
+        )
+
+    return (
+        (circuits_chunk, max(samples_chunk))
+        for circuits_chunk, samples_chunk in zip(
+            _iterate_in_batches(circuits, max_batch_size),
+            _iterate_in_batches(n_samples_per_circuit, max_batch_size),
+        )
+    )
+
+
+def _expand_sample_size(n_samples, max_sample_size):
```

Comment:
Don't you think that having two functions with basically the same name is confusing?

Reply:
Nothing much better comes to my mind – if you come up with something, please change. If not, please [speak the words of wisdom](https://youtu.be/QDYfEBY9NM4?t=31).